File: src/features/browse/redis/clients.ts
// The connected-clients model (M36 §A1) — ported from the prototype's
// `redis-clients.js`, with the mock generator dropped: the rows come from a
// real `CLIENT LIST` via `kvClientList`, so there is nothing to synthesise or
// churn. What stays is everything that gives those rows meaning:
//
// - `INFO_FIELDS` — the field order Redis itself uses, so `infoLine()` rebuilds
//   the exact `CLIENT INFO` line even for a client whose row predates a field.
// - `FIELD_DOC` — one plain sentence per field. A wall of unexplained numbers
//   is what the generic processes tab already was; this is the difference.
// - `KILL_FILTERS` — the six `CLIENT KILL` filters, each rendering the command
//   it will run and matching locally so the tab can show "N clients match"
//   *before* anything is killed.
// - `breakdown()` / `clientRisk()` — the counts the stats strip shows and the
//   four conditions that earn a row attention.
//
// Pure: no React, no invoke. The tab owns the fetching.

import type { KvClient, KvKillFilter } from "./api";

/** The `CLIENT INFO` / `CLIENT LIST` field order, exactly as Redis emits it. */
export const INFO_FIELDS = [
  "id",
  "addr",
  "laddr",
  "fd",
  "name",
  "age",
  "idle",
  "flags",
  "db",
  "sub",
  "psub",
  "ssub",
  "multi",
  "watch",
  "qbuf",
  "qbuf-free",
  "argv-mem",
  "multi-mem",
  "tot-net-in",
  "tot-net-out",
  "rbs",
  "rbp",
  "obl",
  "oll",
  "omem",
  "tot-mem",
  "events",
  "cmd",
  "user",
  "redir",
  "resp",
  "lib-name",
  "lib-ver",
] as const;

/**
 * The client's `CLIENT INFO` line. The server already sent us one (`raw`), so
 * that is what we return — it is authoritative and includes any field this
 * build does not know about. The `INFO_FIELDS` reconstruction is the fallback
 * for a row that somehow arrived without it.
 */
export function infoLine(client: KvClient): string {
  if (client.raw) return client.raw;
  const byField = new Map(client.fields.map((f) => [f.field, f.value]));
  return INFO_FIELDS.map((f) => f + "=" + (byField.get(f) ?? "")).join(" ");
}

/** Read one reported field by name (the inspector renders from these). */
export function clientField(client: KvClient, field: string): string | undefined {
  return client.fields.find((f) => f.field === field)?.value;
}

/**
 * What each `CLIENT LIST` field means, in one sentence. Shown as the hover
 * title of every row in the inspector.
 */
export const FIELD_DOC: Record<string, string> = {
  id: "Unique connection id. Never reused while the server runs — this is what CLIENT KILL ID targets.",
  addr: "Client address and source port, as the server sees it.",
  laddr: "Local address the client connected to. Useful when Redis listens on several interfaces.",
  fd: "File descriptor the connection occupies. A leak shows up here before it shows up in memory.",
  name: "Name the client set with CLIENT SETNAME. Empty means the library never set one.",
  age: "Seconds since the connection opened.",
  idle: "Seconds since the last command. High idle on a normal client usually means a leaked pool connection.",
  flags: "N normal · S replica · M master · b blocked · x in MULTI · O monitor · P pubsub.",
  db: "Currently selected database index.",
  sub: "Channel subscriptions held by this client.",
  psub: "Pattern subscriptions held by this client.",
  ssub: "Shard channel subscriptions held by this client (cluster pub/sub).",
  multi: "Commands queued inside MULTI. -1 means no transaction is open.",
  watch: "Keys being WATCHed for optimistic locking.",
  qbuf: "Query buffer in use, in bytes. A large value means a slow or huge inbound command.",
  "qbuf-free": "Free space left in the query buffer.",
  "argv-mem": "Memory used by the arguments of the command in flight.",
  "multi-mem": "Memory held by the commands queued in an open MULTI.",
  "tot-mem":
    "Total memory this connection costs the server, buffers included. This is the number that matters when clients are the memory problem.",
  "tot-net-in": "Total bytes read from this client.",
  "tot-net-out": "Total bytes written to this client.",
  rbs: "Read buffer size currently allocated for this client.",
  rbp: "Peak position reached in the read buffer.",
  obl: "Bytes queued in the fixed-size output buffer.",
  oll: "Output list length — replies queued but not yet flushed. Growing here means a slow consumer.",
  omem: "Memory of the output buffer. Hits client-output-buffer-limit before anything else.",
  events: "Event loop interest: r readable, w writable.",
  cmd: "Last command executed, including the subcommand.",
  user: "ACL user this connection authenticated as.",
  redir:
    "Client id that receives this connection's tracking invalidations. -1 when tracking is off.",
  resp: "RESP protocol version in use — 2 or 3.",
  "lib-name": "Client library, as reported by CLIENT SETINFO.",
  "lib-ver": "Client library version, as reported by CLIENT SETINFO.",
};

/** One `CLIENT KILL` filter: how it renders, and how it matches locally. */
export interface KillFilterSpec {
  id: KvKillFilter;
  label: string;
  /** The exact command that will run, for the preview. */
  cmd: (value: string) => string;
  /** Placeholder describing what the value is. */
  hint: string;
  /** The same predicate the server applies — powers the live match count. */
  match: (client: KvClient, value: string) => boolean;
}

/**
 * The six filters `CLIENT KILL` supports. Per-row killing does not scale to
 * twenty leaked connections; this is the surface that does.
 */
export const KILL_FILTERS: KillFilterSpec[] = [
  {
    id: "id",
    label: "By ID",
    cmd: (v) => "CLIENT KILL ID " + v,
    hint: "a single connection",
    match: (c, v) => String(c.id) === v.trim(),
  },
  {
    id: "addr",
    label: "By address",
    cmd: (v) => "CLIENT KILL ADDR " + v,
    hint: "one ip:port",
    match: (c, v) => c.addr === v.trim(),
  },
  {
    id: "laddr",
    label: "By local address",
    cmd: (v) => "CLIENT KILL LADDR " + v,
    hint: "everything on one listener",
    match: (c, v) => c.laddr === v.trim(),
  },
  {
    id: "type",
    label: "By type",
    cmd: (v) => "CLIENT KILL TYPE " + v,
    hint: "normal · pubsub · replica",
    match: (c, v) => c.clientType === v.trim(),
  },
  {
    id: "user",
    label: "By ACL user",
    cmd: (v) => "CLIENT KILL USER " + v,
    hint: "all connections of one user",
    match: (c, v) => c.user === v.trim(),
  },
  {
    id: "maxage",
    label: "Older than",
    cmd: (v) => "CLIENT KILL MAXAGE " + v,
    hint: "age in seconds (Redis 7.4+)",
    match: (c, v) => {
      // A blank box must match NOTHING, not everything: `Number("")` is 0, and
      // `age >= 0` would have the popover promise to kill every connection
      // before anything had been typed.
      const raw = v.trim();
      const seconds = Number(raw);
      return raw !== "" && Number.isFinite(seconds) && c.age >= seconds;
    },
  },
];

/** The type colors the segmented control, the stats strip and the row tag share. */
export const CLIENT_TYPE_COLOR: Record<string, string> = {
  normal: "var(--accent)",
  pubsub: "#c792ea",
  replica: "#e2b340",
  master: "#e2b340",
};

/** Seconds a normal client may idle before it looks like a leaked connection. */
export const STALE_IDLE_SECONDS = 300;

/** The counts the stats strip shows, derived from one list — one source of truth. */
export interface ClientBreakdown {
  total: number;
  normal: number;
  pubsub: number;
  replica: number;
  blocked: number;
  /** Normal clients idle over 5 minutes — usually leaked pool connections. */
  stale: number;
  /** Total `tot-mem` across every connection. */
  mem: number;
}

export function breakdown(list: KvClient[]): ClientBreakdown {
  return {
    total: list.length,
    normal: list.filter((c) => c.clientType === "normal").length,
    pubsub: list.filter((c) => c.clientType === "pubsub").length,
    replica: list.filter((c) => c.clientType === "replica" || c.clientType === "master").length,
    blocked: list.filter((c) => c.flags.includes("b")).length,
    stale: list.filter(isStale).length,
    mem: list.reduce((sum, c) => sum + c.totMem, 0),
  };
}

/** A normal client idle past the leak threshold. */
export function isStale(client: KvClient): boolean {
  return client.clientType === "normal" && client.idle > STALE_IDLE_SECONDS;
}

/** Whether the client is parked in a blocking command (`BLPOP`, `XREAD`…). */
export function isBlocked(client: KvClient): boolean {
  return client.flags.includes("b");
}

/** A row that has earned attention, and why. */
export interface ClientRisk {
  sev: "warn" | "note";
  text: string;
}

/**
 * The four things worth flagging on a connection. Your own connection is never
 * flagged — it is doing exactly what you asked it to.
 */
export function clientRisk(client: KvClient): ClientRisk | null {
  if (client.isSelf) return null;
  if (isStale(client)) {
    return {
      sev: "warn",
      text: "Idle " + humanAge(client.idle) + " — likely a leaked pool connection",
    };
  }
  if (client.oll > 6) {
    return {
      sev: "warn",
      text: client.oll + " replies queued — slow consumer, output buffer is filling",
    };
  }
  if (client.multi >= 0 && client.idle > 30) {
    return {
      sev: "warn",
      text: "Transaction open and idle " + humanAge(client.idle) + " — holds WATCHed keys",
    };
  }
  if (client.totMem > 48000) {
    return { sev: "note", text: "Costs " + humanClientMem(client.totMem) + " of server memory" };
  }
  return null;
}

/** A flag letter spelled out (`b` → `blocked`). */
export function flagLabel(flag: string): string {
  const labels: Record<string, string> = {
    N: "normal",
    S: "replica",
    M: "master",
    b: "blocked",
    x: "in MULTI",
    O: "monitor",
    P: "pubsub",
    t: "tracking",
    T: "no-touch",
    e: "no-evict",
  };
  return labels[flag] ?? flag;
}

/** Every flag letter on a client, spelled out (`N` → `N · normal`). */
export function flagSummary(flags: string): string {
  if (!flags) return "—";
  const parts = [...flags].map(flagLabel).filter((label) => label !== "");
  return flags + " · " + parts.join(", ");
}

/** Seconds → `12s` / `4m` / `2h 30m` / `3d 4h`. */
export function humanAge(seconds: number): string {
  if (seconds < 60) return seconds + "s";
  if (seconds < 3600) return Math.floor(seconds / 60) + "m";
  if (seconds < 86400) {
    return Math.floor(seconds / 3600) + "h " + Math.floor((seconds % 3600) / 60) + "m";
  }
  return Math.floor(seconds / 86400) + "d " + Math.floor((seconds % 86400) / 3600) + "h";
}

/** Client memory bytes → `812 B` / `20.5 KB` / `1.20 MB`. */
export function humanClientMem(bytes: number): string {
  if (bytes < 1024) return bytes + " B";
  if (bytes < 1048576) return (bytes / 1024).toFixed(1) + " KB";
  return (bytes / 1048576).toFixed(2) + " MB";
}

/** Network totals → `812 B` / `20 KB` / `1.2 MB` (coarser than memory). */
export function humanNet(bytes: number): string {
  if (bytes < 1024) return bytes + " B";
  if (bytes < 1048576) return (bytes / 1024).toFixed(0) + " KB";
  return (bytes / 1048576).toFixed(1) + " MB";
}

File: src-tauri/src/engines/redis/clients.rs
//! Connected-client administration (M36 §A): `CLIENT LIST` → [`KvClient`], and
//! the three `CLIENT KILL` routes the clients tab offers (one id, a batch of
//! ids, a server-side filter).
//!
//! A `CLIENT LIST` line is exactly the `CLIENT INFO` wire format — space
//! separated `key=value` in the server's own field order. We keep the line
//! verbatim (`raw`) and its pairs in order (`fields`) so the inspector shows
//! what the server actually said, and additionally parse the handful of fields
//! the table sorts, filters and colours by. Nothing is invented: a field the
//! server does not report is simply absent from `fields`.
//!
//! Everything here is server-global (`CLIENT …` ignores the selected db), so
//! db 0 is only the carrier connection. In cluster mode this is inherently
//! **per node** — the node we are attached to.

use async_trait::async_trait;

use crate::shared::error::AppError;
use crate::shared::keyvalue::{ClientAdmin, KvClient, KvField, KvKillFilter};

use super::error::map_query_error;
use super::RedisKvConnection;

#[async_trait]
impl ClientAdmin for RedisKvConnection {
    async fn client_list(&self) -> Result<Vec<KvClient>, AppError> {
        let mut conn = self.conn_for(0).await?;
        let self_id: i64 = redis::cmd("CLIENT")
            .arg("ID")
            .query_async(&mut conn)
            .await
            .map_err(map_query_error)?;
        let list: String = redis::cmd("CLIENT")
            .arg("LIST")
            .query_async(&mut conn)
            .await
            .map_err(map_query_error)?;
        Ok(parse_client_list(&list, self_id))
    }

    async fn client_kill(&self, filter: KvKillFilter, value: &str) -> Result<u64, AppError> {
        if value.trim().is_empty() {
            return Err(AppError::Invalid(
                "a CLIENT KILL filter needs a value to match on.".into(),
            ));
        }
        let mut conn = self.conn_for(0).await?;
        kill_one(&mut conn, filter.as_token(), value).await
    }

    async fn client_kill_ids(&self, ids: &[i64]) -> Result<u64, AppError> {
        let mut conn = self.conn_for(0).await?;
        let mut closed = 0;
        // One command per id — the same list of `CLIENT KILL ID <id>` lines the
        // confirm dialog showed. Multi-id kills are not portable across the
        // server versions ByteTable supports.
        for id in ids {
            closed += kill_one(&mut conn, "ID", &id.to_string()).await?;
        }
        Ok(closed)
    }

    async fn client_no_evict(&self, on: bool) -> Result<(), AppError> {
        let mut conn = self.conn_for(0).await?;
        redis::cmd("CLIENT")
            .arg("NO-EVICT")
            .arg(if on { "on" } else { "off" })
            .query_async::<redis::Value>(&mut conn)
            .await
            .map_err(map_query_error)?;
        Ok(())
    }

    async fn client_unpause(&self) -> Result<(), AppError> {
        let mut conn = self.conn_for(0).await?;
        redis::cmd("CLIENT")
            .arg("UNPAUSE")
            .query_async::<redis::Value>(&mut conn)
            .await
            .map_err(map_query_error)?;
        Ok(())
    }
}

/// Run one `CLIENT KILL <filter> <value>` and read the "closed" count. The
/// filter form always replies with an integer (the legacy `addr:port` form
/// replies `+OK`, which is why we never use it).
async fn kill_one(
    conn: &mut redis::aio::MultiplexedConnection,
    filter: &str,
    value: &str,
) -> Result<u64, AppError> {
    let killed: i64 = redis::cmd("CLIENT")
        .arg("KILL")
        .arg(filter)
        .arg(value)
        .query_async(conn)
        .await
        .map_err(map_query_error)?;
    Ok(killed.max(0) as u64)
}

/// Parse the newline-separated `CLIENT LIST` reply into [`KvClient`] rows.
pub(super) fn parse_client_list(reply: &str, self_id: i64) -> Vec<KvClient> {
    reply
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(|line| parse_client_line(line, self_id))
        .collect()
}

/// Parse one `CLIENT LIST` / `CLIENT INFO` line.
fn parse_client_line(line: &str, self_id: i64) -> KvClient {
    // Keep every pair in the server's order — that order IS the wire format.
    let fields: Vec<KvField> = line
        .split(' ')
        .filter(|token| !token.is_empty())
        .filter_map(|token| {
            let (field, value) = token.split_once('=')?;
            Some(KvField {
                field: field.to_string(),
                value: value.to_string(),
            })
        })
        .collect();

    let get = |key: &str| {
        fields
            .iter()
            .find(|f| f.field == key)
            .map(|f| f.value.as_str())
    };
    let num = |key: &str, default: i64| get(key).and_then(|v| v.parse().ok()).unwrap_or(default);

    let flags = get("flags").unwrap_or_default().to_string();
    let cmd = get("cmd").unwrap_or_default().to_string();
    let id = num("id", 0);

    KvClient {
        id,
        addr: get("addr").unwrap_or_default().to_string(),
        laddr: get("laddr").unwrap_or_default().to_string(),
        name: get("name").unwrap_or_default().to_string(),
        age: num("age", 0),
        idle: num("idle", 0),
        client_type: client_type(&flags, &cmd),
        flags,
        db: num("db", 0).clamp(0, 255) as u8,
        sub: num("sub", 0),
        psub: num("psub", 0),
        // `multi=-1` is Redis's own "no transaction open" sentinel.
        multi: num("multi", -1),
        watch: num("watch", 0),
        qbuf: num("qbuf", 0),
        oll: num("oll", 0),
        omem: num("omem", 0),
        tot_mem: num("tot-mem", 0),
        cmd,
        user: get("user").unwrap_or_default().to_string(),
        is_self: id == self_id,
        fields,
        raw: line.to_string(),
    }
}

/// The client's class, derived the way Redis's own `CLIENT KILL TYPE` groups
/// them: replica/master connections by flag, pub/sub by the `P` flag or a
/// held subscription, everything else normal.
fn client_type(flags: &str, cmd: &str) -> String {
    if flags.contains('S') {
        return "replica".into();
    }
    if flags.contains('M') {
        return "master".into();
    }
    if flags.contains('P') || cmd.split('|').next().unwrap_or("").ends_with("subscribe") {
        return "pubsub".into();
    }
    "normal".into()
}

#[cfg(test)]
mod tests {
    use super::*;

    const LINE: &str = "id=6 addr=10.0.4.18:52310 laddr=10.0.1.5:6379 fd=8 name=web-1 age=42 \
        idle=0 flags=N db=2 sub=0 psub=0 ssub=0 multi=-1 watch=0 qbuf=26 qbuf-free=20448 \
        argv-mem=10 multi-mem=0 tot-net-in=140 tot-net-out=0 rbs=1024 rbp=0 obl=0 oll=0 omem=0 \
        tot-mem=20512 events=r cmd=client|list user=default redir=-1 resp=3 lib-name=redis-py \
        lib-ver=5.0.4";

    #[test]
    fn parses_every_field_in_server_order_and_flags_self() {
        let clients = parse_client_list(LINE, 6);
        assert_eq!(clients.len(), 1);
        let c = &clients[0];
        assert_eq!(c.id, 6);
        assert_eq!(c.addr, "10.0.4.18:52310");
        assert_eq!(c.laddr, "10.0.1.5:6379");
        assert_eq!(c.name, "web-1");
        assert_eq!(c.age, 42);
        assert_eq!(c.db, 2);
        assert_eq!(c.multi, -1);
        assert_eq!(c.tot_mem, 20512);
        assert_eq!(c.cmd, "client|list");
        assert_eq!(c.user, "default");
        assert_eq!(c.client_type, "normal");
        assert!(c.is_self);
        // The pairs keep the server's order, first and last included.
        assert_eq!(c.fields.first().map(|f| f.field.as_str()), Some("id"));
        assert_eq!(c.fields.last().map(|f| f.field.as_str()), Some("lib-ver"));
        assert_eq!(c.fields.len(), 33);
        // …and the raw line round-trips verbatim (this is CLIENT INFO output).
        assert_eq!(c.raw, LINE);
    }

    #[test]
    fn derives_the_client_type_the_way_client_kill_type_groups_them() {
        assert_eq!(client_type("N", "get"), "normal");
        assert_eq!(client_type("S", "psync"), "replica");
        assert_eq!(client_type("M", "ping"), "master");
        assert_eq!(client_type("P", "ping"), "pubsub");
        // A subscribed client before the server sets the P flag.
        assert_eq!(client_type("N", "psubscribe"), "pubsub");
    }

    #[test]
    fn missing_fields_fall_back_without_inventing_values() {
        let clients = parse_client_list("id=9 addr=x:1 flags=b cmd=blpop", 1);
        let c = &clients[0];
        assert_eq!(c.id, 9);
        assert_eq!(c.name, "");
        assert_eq!(c.idle, 0);
        assert_eq!(c.multi, -1);
        assert!(!c.is_self);
        assert_eq!(c.fields.len(), 4);
    }
}
